model_bump.go

```diff
@@ -3,27 +3,166 @@ package bux
 import (
 	"bytes"
 	"database/sql/driver"
+	"encoding/hex"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"reflect"
+	"sort"
+
+	"github.com/libsv/go-bt/v2"
 )
 
-// BUMP represents BUMP format
+// BUMPPaths represents a slice of BUMPs (BSV Unified Merkle Paths)
+type BUMPPaths []BUMP
+
+// BUMP represents BUMP (BSV Unified Merkle Path) format
 type BUMP struct {
-	BlockHeight uint64         `json:"blockHeight,string"`
-	Path        []BUMPPathMap `json:"path"`
+	BlockHeight uint64       `json:"blockHeight,string"`
+	Path        [][]BUMPNode `json:"path"`
+	// private field for storing already used offsets to avoid duplicate nodes
+	allNodes []map[uint64]bool
 }
 
-// BUMPPathMap represents map with pathes
-type BUMPPathMap map[string]BUMPPathElement
-
-// BUMPPathElement represents each BUMP path element
-type BUMPPathElement struct {
-	Hash      string `json:"hash,omitempty"`
+// BUMPLeaf represents each BUMP path element
+type BUMPNode struct {
+	Offset    uint64 `json:"offset,string"`
+	Hash      string `json:"hash"`
 	TxId      bool   `json:"txid,omitempty"`
 	Duplicate bool   `json:"duplicate,omitempty"`
 }
 
+// CalculateMergedBUMP calculates Merged BUMP from a slice of Merkle Proofs
+func CalculateMergedBUMP(mp []MerkleProof) (BUMP, error) {
+	bump := BUMP{}
+
+	if len(mp) == 0 || mp == nil {
+		return bump, nil
+	}
+
+	height := len(mp[0].Nodes)
+	if height > maxCmpHeight {
+		return bump,
+			fmt.Errorf("BUMP cannot be higher than %d", maxCmpHeight)
+	}
+
+	for _, m := range mp {
+		if height != len(m.Nodes) {
+			return bump,
+				errors.New("Merged BUMP cannot be obtained from Merkle Proofs of different heights")
+		}
+	}
+
+	bump.Path = make([][]BUMPNode, height)
+	bump.allNodes = make([]map[uint64]bool, height)
+	for i := range bump.allNodes {
+		bump.allNodes[i] = make(map[uint64]bool, 0)
+	}
+
+	for _, m := range mp {
+		bumpToAdd := m.ToBUMP()
+		err := bump.add(bumpToAdd)
+		if err != nil {
+			return BUMP{}, err
+		}
+	}
+
+	for _, p := range bump.Path {
+		sort.Slice(p, func(i, j int) bool {
+			return p[i].Offset < p[j].Offset
+		})
+	}
+
+	return bump, nil
+}
+
+func (bump *BUMP) add(b BUMP) error {
+	if len(bump.Path) != len(b.Path) {
+		return errors.New("BUMPs with different heights cannot be merged")
+	}
+
+	for i := range b.Path {
+		for _, v := range b.Path[i] {
+			_, value := bump.allNodes[i][v.Offset]
+			if !value {
+				bump.Path[i] = append(bump.Path[i], v)
+				bump.allNodes[i][v.Offset] = true
+				continue
+			}
+			if i == 0 && value && v.TxId {
+				for j := range bump.Path[i] {
+					if bump.Path[i][j].Offset == v.Offset {
+						bump.Path[i][j] = v
+					}
+				}
+			}
```

Comment:
This section solves a case, where we have two sibling transactions that are client txids. Without this if-case, the field `txId` will not have value `true` for the second sibling that's being added, and therefore, the hex string will be wrong, because the flag will mismatch.

If you can think of a better, more elegant solution to achieve the same, please do let me know.